### PR TITLE
feat(transformer): add decorators plugin

### PR DESCRIPTION
**Pull request by Dunqing**

### CODE REVIEW

**Comment by Dunqing** [comment]
Current dependencies on/for this PR:
* `main`
  * **PR #2139** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/2139?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/oxc-project/oxc/2139?utm_source=stack-comment).

File: crates/oxc_transformer/src/proposals/decorators.rs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        options.decorators.map(|options| Self {
+            ast,
+            _ctx: ctx,
+            options,
+            top_statements: Vec::new(),
+        })
+    }
+
+    pub fn transform_program(&mut self, program: &mut Program<'a>) {
+        program.body.splice(0..0, self.top_statements.drain(..));
```
Comment (magic-akari) [inline crates/oxc_transformer/src/proposals/decorators.rs:60]: Should we consider directives like "use strict"?

Comment (Boshen) [reply crates/oxc_transformer/src/proposals/decorators.rs:60]: directives are in another field for the exact situations like this ;-)

https://github.com/oxc-project/oxc/blob/fda2d7c334b0ec5dd57656fd8ae1601460594af2/crates/oxc_ast/src/ast/js.rs#L24

Comment (Boshen) [reply crates/oxc_transformer/src/proposals/decorators.rs:60]: I don't get your questions.

`super` is a special expression 🤔 

https://github.com/oxc-project/oxc/blob/fda2d7c334b0ec5dd57656fd8ae1601460594af2/crates/oxc_ast/src/ast/js.rs#L56

**Comment by Boshen** [comment]
### Merge activity

* **Jan 23, 10:09 PM**: @Boshen started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/oxc-project/oxc/2139).

**Pull request merged**